lib/Dialect/ESI/runtime/python/esiaccel/cosim/simulator.py

```diff
@@ -93,11 +94,66 @@ class Simulator:
   # broken behavior by overriding this.
   UsesStderr = True
 
-  def __init__(self, sources: SourceFiles, run_dir: Path, debug: bool):
+  def __init__(self,
+               sources: SourceFiles,
+               run_dir: Path,
+               debug: bool,
+               run_stdout_callback: Optional[Callable[[str], None]] = None,
+               run_stderr_callback: Optional[Callable[[str], None]] = None,
+               compile_stdout_callback: Optional[Callable[[str], None]] = None,
+               compile_stderr_callback: Optional[Callable[[str], None]] = None,
+               make_default_logs: bool = True):
+    """Simulator base class.
+
+    Optional sinks can be provided for capturing output. If not provided,
+    the simulator will write to log files in `run_dir`.
+
+    Args:
+      sources: SourceFiles describing RTL/DPI inputs.
+      run_dir: Directory where build/run artifacts are placed.
+      debug: Enable cosim debug mode.
+      run_stdout_callback: Line-based callback for runtime stdout.
+      run_stderr_callback: Line-based callback for runtime stderr.
+      compile_stdout_callback: Line-based callback for compile stdout.
+      compile_stderr_callback: Line-based callback for compile stderr.
+      make_default_logs: If True and corresponding callback is not supplied,
+        create log file and emit via internally-created callback.
+    """
     self.sources = sources
     self.run_dir = run_dir
     self.debug = debug
 
+    # Install callbacks (possibly wrapping default file log writers).
+    self._run_stdout_cb = run_stdout_callback
+    self._run_stderr_cb = run_stderr_callback
+    self._compile_stdout_cb = compile_stdout_callback
+    self._compile_stderr_cb = compile_stderr_callback
+
+    self._default_files: List[IO[str]] = []
+
+    def _ensure_default_log(cb_attr: str, log_attr: str, filename: str):

```

Comment:
I don't like the usage of `setattr` and `getattr` here. It's something to be avoided unless it's necessary or saves a lot of code.

Suggested refactor:
`(self._run_stdout_cb, self._run_stdout_log) = ensure_default(run_stdout_callback, 'sim_stdout.log')`

and necessary changes to `_ensure_default`.